### ISSUE TITLE
SpinLockDelay cause performance slow

### ISSUE THREAD

**Issue by alk** [gperftools/gperftools#663]
Originally reported on Google Code with ID 660
```
What steps will reproduce the problem?
1. on CentOS 4.3
2. kernel is
   2.6.32_1-14-0-0 #1 SMP Mon Mar 31 10:42:09 CST 2014 x86_64 x86_64 x86_64 GNU/Linux
3. libc is
GNU C Library stable release version 2.3.4
Compiled by GNU CC version 3.4.4 20050721 (Red Hat 3.4.4-2).
Compiled on a Linux 2.4.20 system on 2006-03-08.
4. tcmalloc version gperftools-2.1

I'm doing performance test. I see that the performance becoming slower and slower when
using tcmalloc, if not using tcmalloc, the the performance is good and stable.

gstack shows a lot of threads(48) blocked on SpinLockDelay

stack is like this
Thread 51 (Thread 1895635296 (LWP 27838)):
#0  0x00000000008205b6 in base::internal::SpinLockDelay ()
#1  0x000000000082043b in SpinLock::SlowLock ()
#2  0x0000000000817630 in tcmalloc::CentralFreeList::InsertRange ()
#3  0x000000000081f7a1 in tcmalloc::ThreadCache::ReleaseToCentralCache ()
#4  0x000000000081f9ac in tcmalloc::ThreadCache::Scavenge ()
#5  0x00000000008348ce in tc_delete ()
#6  0x000000302d390562 in std::string::_Rep::_M_destroy ()
#7  0x000000302d3907a8 in std::string::reserve ()
#8  0x000000302d39102c in std::string::append () from /usr/lib64/libstdc++.so.6
#9  0x00000000005a95c2 in Json::FastWriter::writeValue ()
#10 0x00000000005a981d in Json::FastWriter::writeValue ()
#11 0x00000000005a981d in Json::FastWriter::writeValue ()
#12 0x00000000005a981d in Json::FastWriter::writeValue ()
#13 0x00000000005a981d in Json::FastWriter::writeValue ()
#14 0x00000000005a939b in Json::FastWriter::write ()
.....

I think this maybe because my libc version is too low?

please help!
Thanks a lot!


```

Reported by `zlxderek` on 2014-12-04 08:54:50

**Comment by alk** [comment]
Closing as duplicate of O(N) in page heap
